control/server.py

```diff
@@ -45,6 +47,18 @@ def sigchld_handler(signum, frame):
     # GW process should exit now
     raise SystemExit(f"Gateway subprocess terminated {pid=} {exit_code=}")
 
+class RPCGuard:
+    def __init__(self, timeout = None) -> None:
+        self.rpc_lock = threading.RLock()
+        self.lock_timeout = timeout
+
+    def __enter__(self):
+        self.rpc_lock.acquire(True, self.lock_timeout)
```

Comment:
@baum I used a timeout so we wouldn't end up stuck in a deadlock. I used really long one so we should only timeout in case of some disaster. I didn't add code to handle such case as it's not such a big deal if it happens. It just means that we reverted to the logic that we have right now. It doesn't mean that we'll get the actual problem. And even if we do it's not that damaging.

Reply:
How we can reach a deadlock here?